### PR TITLE
Weighted A*

### PR DESCRIPTION
**Pull request by RoosterDragon**
Replace Constants.CellCost and Constants.DiagonalCellCost with a dynamically calculated value based on the lowest cost terrain to traverse. Using a fixed value meant the pathfinder heuristics would be incorrect.

In the four default mods, the minimum cost is in fact 100, not 125. This increase would essentially allow the pathfinder to return suboptimal paths up to 25% longer in the worst case, but it would be quicker to do so.

This is exactly what Weighted A* does - overestimate the heuristic by some factor in order to speed up the search by checking fewer routes. This makes the heuristic inadmissible and it may now return suboptimal paths, but their worst case length is bounded by the weight. A weight of 125% will never produce paths more than 25% longer than the shortest, optimal, path.

We set the default weight to 25% to effectively maintain the existing, suboptimal, behaviour due to the choice of the old constant - in future it may prove a useful tuning knob for performance.

----

More reading if you are looking for more in depth on what this is and think terms like "ε-admissible" are cool:

[Wikipedia on A*](https://en.wikipedia.org/wiki/A*_search_algorithm) - specifically the section on [Bounded relaxation](https://en.wikipedia.org/wiki/A*_search_algorithm#Bounded_relaxation) 

[A* and Weighted A* Search](https://www.cs.cmu.edu/~motionplanning/lecture/Asearch_v8.pdf)

----

Below are some figures from moving a lone rifle between opposite corners (TL/BR) of the 128x128 RA map, Pool Party.

The path search runs two instances until they meet in the middle. The nodes considered in the search effectively determine how much work the search has performed. This gives an impression idea of the tradeoff of allowing suboptimal paths.

This isn't an extensive performance test obviously, but is just here to give a ballpark impression. I figured the best scenario was to maintain the existing weight (even if it was likely unintended) and now the tuning knob is easily accessible somebody could muck about with it in future if they wish and are willing to investigate the effects proper.

Weight | Considered | Total | Considered %
-- | -- | -- | --
100 | 5093+5091 | 10184 | 100%
105 | 4777+4793 | 9570 | 94%
110 | 4613+4554 | 9167 | 90%
115 | 4375+4353 | 8728 | 86%
120 | 4091+3999 | 8090 | 79%
125 | 3607+3586 | 7193 | 71%
130 | 2644+2670 | 5314 | 52%
135 | 2347+2311 | 4658 | 46%
140 | 1929+1914 | 3843 | 38%
150 | 1609+1679 | 3288 | 32%
175 | 907+1078 | 1985 | 19%
200 | 663+854 | 1517 | 15%
250 | 542+709 | 1251 | 12%
300 | 512+665 | 1177 | 12%
500 | 608+630 | 1238 | 12%

### CODE REVIEW

File: OpenRA.Mods.Common/Pathfinder/PathSearch.cs [hunk 1]
```diff
@@ -57,40 +57,33 @@ public static IPathSearch Search(World world, Locomotor locomotor, Actor self, B
...
+			search.heuristic = search.DefaultEstimator(target);
+
+			// The search will aim for the shortest path by default, a weight of 100%.
+			// We can allow the search to find paths that aren't optimal by changing the weight.
+			// We provide a weight that limits the worst case length of the path,
+			// e.g. a weight of 110% will find a path no more than 10% longer than the shortest possible.
+			// The benefit of allowing the search to return suboptimal paths is faster computation time.
+			// The search can skip some areas of the search space, meaning it has less work to do.
+			// We allow paths up to 25% longer than the shortest, optimal path, to improve pathfinding time.
+			search.heuristicWeightPercentage = 125;
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Pathfinder/PathSearch.cs]: Would using `.WithHeuristicWeight` override this?

Comment (RoosterDragon) [reply OpenRA.Mods.Common/Pathfinder/PathSearch.cs]: Yes.

**Review by reaperrr — APPROVED** [review]
Code looks good to me and didn't encounter any issues while testing

**Comment by reaperrr** [comment]
During my local tests, a value of 150% didn't produce notably longer paths, while notably reducing the performance hit, so that value looks like a good compromise for a follow-up.

I know @teinarss is working on a larger pathfinding refactor, but I think in the mean time we should take this + a follow-up that increases the default to ~150% to improve pathfinding performance for Next+1.

**Comment by reaperrr** [comment]
@teinarss @tovl any objections to merging this?

File: OpenRA.Mods.Common/Pathfinder/PathSearch.cs [hunk 2]
```diff
@@ -57,40 +57,33 @@ public static IPathSearch Search(World world, Locomotor locomotor, Actor self, B
 
 		public static IPathSearch FromPoint(World world, Locomotor locomotor, Actor self, CPos @from, CPos target, BlockedByActor check)
```
Comment (tovl) [inline OpenRA.Mods.Common/Pathfinder/PathSearch.cs]: `FromPoint` seems to be a bit pointless now that it is basically just directing to `FromPoints` without any logic of it's own. Would there be any objection to replacing all references to `FromPoint` with `FromPoints` and removing it?

Comment (teinarss) [reply OpenRA.Mods.Common/Pathfinder/PathSearch.cs]: It harbors the important logic of creating the `froms` array! :wink:

**Review by teinarss — APPROVED** [review]

File: OpenRA.Mods.Common/Pathfinder/PathSearch.cs (hunk 2 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Pathfinder/PathSearch.cs]: If we dropped this we'd have to create a single-entry array in every place that currently uses `FromPoint`, which is in half a dozen places spread over 5 files.
Might be worth considering, but not a blocker for this PR, in my opinion.

**Pull request merged**